pkg/backporter/backporter.go

```diff
@@ -0,0 +1,292 @@
+	emptyTemplate  = template.Must(template.New("empty").Parse("{{.}}"))
+)
+
+// HandlerFuncWithErrorReturn allows returning errors to be logged
+type HandlerFuncWithErrorReturn func(http.ResponseWriter, *http.Request) error
+
+// Class to hold the UI data for the clones page
+type clonesTemplateData struct {
+	Bug    *bugzilla.Bug          //bug details
+	Clones []*bugzilla.Bug        // List of clones for the bug
+	Parent *bugzilla.Bug          // Root bug if it is a a bug, otherwise holds itself
+	PRs    []bugzilla.ExternalBug // Details of linked PR
+}
+
+//Writes an HTML page, prepends header in htmlPageStart and appends header from htmlPageEnd around tConstructor.
+func writePage(w http.ResponseWriter, title string, body *template.Template, data interface{}) error {
+	fmt.Fprintf(w, htmlPageStart, title)
+
+	if err := body.Execute(w, data); err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "%s: %v", http.StatusText(http.StatusInternalServerError), err)
+		return err
+	}
+	fmt.Fprint(w, htmlPageEnd)
+	return nil
+}
+
+// GetLandingHandler will return a simple bug search page
+func GetLandingHandler() HandlerFuncWithErrorReturn {
+	return func(w http.ResponseWriter, req *http.Request) error {
+		writePage(w, "Home", emptyTemplate, nil)
+		return nil
+	}
+}
+
+// GetBugHandler returns a function with bug details  in JSON format
+func GetBugHandler(client bugzilla.Client) HandlerFuncWithErrorReturn {
+	return func(w http.ResponseWriter, r *http.Request) error {
+		if r.Method != "GET" {
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte(http.StatusText(http.StatusBadRequest)))
+			return fmt.Errorf("Not a GET request")
+		}
+		bugIDStr := r.URL.Query().Get(BugIDQuery)
+		bugID, err := strconv.Atoi(bugIDStr)
+		if bugIDStr == "" || err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			fmt.Fprintf(w, "%s query missing or incorrect", BugIDQuery)
+			return err
+		}
+		bugInfo, err := client.GetBug(bugID)
+		if err != nil {
+			w.WriteHeader(http.StatusNotFound)
+			w.Write([]byte("Bug ID not found"))
+			return err
+		}
+		jsonBugInfo, err := json.MarshalIndent(*bugInfo, "", "  ")
+		if err != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, "failed to marshal bugInfo to JSON: %v", err)
+			// logger.WithError(err).Errorf("failed to marshal bugInfo to JSON")
+			return err
+		}
+		w.WriteHeader(http.StatusOK)
+		w.Write(jsonBugInfo)
+		return nil
+	}
+}
+
+// GetClonesHandler returns an HTML page with detais about the bug and its clones
+func GetClonesHandler(client bugzilla.Client) HandlerFuncWithErrorReturn {
+	return func(w http.ResponseWriter, req *http.Request) error {
+		bugIDStr := req.URL.Query().Get(BugIDQuery)
+		bugID, err := strconv.Atoi(bugIDStr)
+		if bugIDStr == "" || err != nil {
+			w.WriteHeader(http.StatusBadRequest)
+			writePage(w, "Error!", emptyTemplate, fmt.Sprintf("%s - query incorrect", BugIDQuery))
```

Comment:
Nit: Would be nice to return the error

Reply:
It's not just a nit, the error handling could be better overall: `writePage` returns an error but it's basically never read. Either that method should not return an error (it could just log problems), or its errors should not be ignored.

Reply:
@petr-muller The error returned by error handler is meant to be logged in the handleWithMetrics function in the recordError function. Is that not a good way to log errors?

Reply:
What I meant was the `writePage`'s return value, which is not checked anywhere. So either that function should not return errors at all (and handle them locally, by e.g. just logging), or it should be checked and handled at the callsites. But having `writePage` return errors and ignoring them at callsites seems wrong.

Reply:
> Nit: Would be nice to return the error

@alvaroaleman I was returning only the error for the empty string, now I have separated both of those conditions out

Reply:
@petr-muller Yup. got what you meant now. I'll make sure the returned errors are being used for writePage